common/src/main/java/me/shedaniel/clothconfig2/api/ValueHolder.java

```diff
@@ -0,0 +1,7 @@
+package me.shedaniel.clothconfig2.api;
+
+import org.jetbrains.annotations.Nullable;
+
+public interface ValueHolder<T> {
+    @Nullable T getValue();
```

Comment:
It'd be great if this could be `NotNull`, that'd need changes in `EmptyEntry` for example though. As it stands, there's several unchecked dereferences that could cause NPEs, for example in the `Requirement` factory methods.

Reply:
I think it is a valid use case for value to be null, so no

Reply:
OK, in that case I'll have to fix any potential NPEs before merging :+1:

Reply:
I've fixed the NPE issues in 688d590bfcb0ef62817795d747985167a4fbaea7 and also proposed a slightly different interface, where specific Config Entry classes can decide from themselves if their value is nullable (69d3590d2fa93476f1c9309b2ba93a398ca7eff3)

Reply:
I believe that kind of makes it more annoying to deal with, I think one single interface where it is always nullable should be fine. Especially as the implementers should already know if they are dealing with nullable values. I am not sure what a NotNullValueHolder would benefit.

Reply:
Quoting JetBrains' guide:

> Overriding/implementing methods with an annotated declaration:
>
> - The `@NotNull` annotation of the parent method requires the `@NotNull` annotation for the child class method.
> - Methods with the `@Nullable` annotation in the parent method can have either `@Nullable` or `@Notaull` annotations in the child class method.

So there isn't any need for multiple interfaces, the relevant classes could simply declare the `@NotNull` annotation on their implementation.

The main justification is to make it explicit to users and IDEs that their "clean" code is actually safe without null checks. Otherwise IDEs will complain about not checking nullable values.

Adding those annotations could be part of a separate PR though

Reply:
I don't really want a spam of @NotNull annotations, I used to add them all over the place, but stopped doing it now, so you don't really see them much in newer code.
I like the idea that everything is implicitly non-null, where @Nullable is declared on nullable values. I don't really care about IDEs or even Kotlin for their strict nullability checks, it is way too much maintenance hassle for a simple thing.

Reply:
I've reverted back to the original `ValueHolder`. Kept the cleaner demo code, even though it technically violates `@Nullable`